### ISSUE TITLE
TextField announces value changes after it loses focus

### ISSUE THREAD

**Issue by alanrussian** [flutter/flutter#23180]
I have a screen where there's a slider and a text box where both are in sync. When the user changes the value in the text box, I want the text box to announce the change; and when the user changes the value in the slider, I want the slider to announce the change. It looks something like this:

![image](https://user-images.githubusercontent.com/2225255/47052436-294ee480-d15d-11e8-8434-0ffad4e8b56b.png)

There is a problem in the following scenario:

0. Turn on TalkBack.
1. Focus on Slider and move it using the volume keys. You hear "Fifty percent seek control"
2. Focus on Text Field. Double tap to bring up the keyboard. Then dismiss the keyboard such that the text field loses focus.
3. Focus on Slider. It's still at 50%. Move the slider by pressing the volume up key. You hear "Fifty-five percent seek control. Five replaced zero."

The problem is that the text field is still announcing changes to its contents after it loses focus. This only happens after focusing on the text field, so I feel like this is a bug.

Code of sample app demonstrating this problem:

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  final String title;

  const MyHomePage({Key key, this.title}) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  double _value = 0.0;
  TextEditingController _controller;

  @override
  void initState() {
    super.initState();
    _controller = TextEditingController(text: '0');
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Column(
        children: <Widget>[
          TextField(
            controller: _controller,
          ),
          Slider(
            onChanged: (value) {
              setState(() {
                _value = value;
              });
              _controller.text = (value * 100).round().toString();
            },
            value: _value,
          )
        ],
      ),
    );
  }
}
```

**Comment by jamesderlin** [comment]
CC @jonahwilliams @goderbauer

**Comment by jonahwilliams** [comment]
Even if I type numbers in without TalkBack on, the slider does not adjust - you're missing a listener in there for the text field.

**Comment by alanrussian** [comment]
That direction seems to work fine, and I'm not worried about it. I just didn't implement that direction in my simple reproduction case.

**Comment by alanrussian** [comment]
@jonahwilliams just wanted to confirm you're not waiting on anything from me for this bug. Thanks!

**Comment by jonahwilliams** [comment]
Sorry, lost track of it.

The TextField needs to announce values not just when it has a11y focus, but input focus. This is because interacting with an onscreen keyboard will move a11y focus to the last pressed key. I think in this case we need the slider to take focus from the text field when it's interacted with via a FocusNode. This might be a more general problem since IIRC only text fields take focus, possibly part of https://github.com/flutter/flutter/issues/13264

**Comment by alanrussian** [comment]
Np. I think I'm noticing that the value changes are being announced even after the text field has lost input focus. In the application I'm working on (not the simple repro case above), we have a FocusNode and we do stuff when the focus is lost. I'm noticing that the FocusNode value is false but the text changes are still being announced.

I just noticed another thing. Remember the instructions from above:

1. Turn on TalkBack.
2. Focus on Slider and move it using the volume keys. You hear "Fifty percent seek control"
3. Focus on Text Field. Double tap to bring up the keyboard. Then dismiss the keyboard such that the text field loses focus.
4. Focus on Slider. It's still at 50%. Move the slider by pressing the volume up key. You hear "Fifty-five percent seek control. Five replaced zero."

If I swap 1 and 3, I can't reproduce the issue:

1. Focus on Slider and move it using the volume keys. You hear "Fifty percent seek control"
2. Focus on Text Field. Double tap to bring up the keyboard. Then dismiss the keyboard such that the text field loses focus.
3. Turn on TalkBack.
4. Focus on Slider. It's still at 50%. Move the slider by pressing the volume up key. You hear "Fifty-five percent seek control. Five replaced zero."

**Comment by jonahwilliams** [comment]
I'll definitely do some more investigation tomorrow. It may turn out that we've got some stale information being propagated through the semantics system in the case of input focus. Stay tuned.

**Comment by jonahwilliams** [comment]
Update: I've found the issue. When creating the text field semantics on the android a11y bridge we don't correctly account for a fields value being changed even if it isn't focused. It is currently unconditionally set to a live region. It should only be set if it has input focus

**Comment by jonahwilliams** [comment]
For reference the sample code I am working with:

Focusing the second text field causes the first the stop making the announcements. Focusing the slider without unfocusing the first causes the announcements to persist.

```dart
class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  final String title;

  const MyHomePage({Key key, this.title}) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  double _value = 0.0;
  TextEditingController _controller;
  @override
  void initState() {
    super.initState();
    _controller = TextEditingController(text: '0');
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Column(
        children: <Widget>[
          TextField(
            controller: _controller,
            onChanged: (String newValue) {
              setState(() {
                if (newValue == '') {
                  _value = 0;
                } else {
                  _value = double.tryParse(newValue);
                }
              });
            },
          ),
          TextField(
            controller: TextEditingController(text: 'asd'),
            onChanged: (String newValue) {
              setState(() {
                if (newValue == '') {
                  _value = 0;
                } else {
                  _value = double.tryParse(newValue);
                }
              });
            },
          ),
          Slider(
            onChanged: (value) {
              setState(() {
                _value = value;
              });
              _controller.text = value.round().toString();
            },
            min: 0.0,
            max: 100.0,
            value: _value,
          )
        ],
      ),
    );
  }
}
```

**Comment by alanrussian** [comment]
@jonahwilliams I'm still noticing this issue in our app

**Comment by jonahwilliams** [comment]
are you explicitly releasing the focus of the text field? This isn't done automatically per the reasoning above

**Comment by alanrussian** [comment]
The text field does not have input focus. I'm tapping the bottom right button on the keyboard which is closing the keyboard and removing the focus. The focus node says there is no focus.

**Comment by dnfield** [comment]
@alanrussian Can you confirm whether this is Android, iOS, or both?

**Comment by alanrussian** [comment]
I have verified on Android not iOS. I can test on iOS if you need.

**Comment by jonahwilliams** [comment]
iOS doesn't announce things, so it would only occur on android

**Comment by dnfield** [comment]
Good enough, thanks.  Just wanted to make sure I'm chasing on the right platform.

**Comment by dnfield** [comment]
Root cause:

- We announce things form text fields when they have input focus even if they don't have a11y focus because we're assuming the keyboard can take a11y focus away but we still want to announce the text field.
- We don't unfocus text fields when other elements get focused by default - but this is consistent with the platform side.

I'm looking to see if I can change this so that text fields check if keyboard has a11y focus or not.  In the interim, you could work around this by manually unfocusing the TextField like so:

```dart
@override
  Widget build(BuildContext context) {
    FocusNode focusNode = FocusNode();
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Column(
        children: <Widget>[
          TextField(
            controller: _controller,
            focusNode: focusNode,
          ),
          Slider(
            onChanged: (value) {
              setState(() {
                _value = value;
              });
              _controller.text = (value * 100).round().toString();
              focusNode.unfocus();
            },
            value: _value,
          )
        ],
      ),
    );
  }
```

**Comment by dnfield** [comment]
Ahh I see now.  I was having trouble reproducing that, but now I can.
